### PR TITLE
enh_T1_analysis tool

### PR DESCRIPTION
**Pull request by ByoungMooAnn**
Changes proposed in this pull request:
- enh T1 analysis tool included.

### CODE REVIEW

File: pycqed/analysis_v2/Two_state_T1_analysis.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+                 ):
+        if options_dict is None:
+            options_dict = dict()
+        super().__init__(t_start=t_start, t_stop=t_stop, label=label,
+                         options_dict=options_dict, close_figs=close_figs,
+                         do_fitting=True)
+        # used to determine how to determine 2nd excited state population
+        self.classification_method = classification_method
+        self.rates_ch_idx = rates_ch_idx
+        self.d1 = 2
```
Comment (AdriaanRol) [inline pycqed/analysis_v2/Two_state_T1_analysis.py]: Assuming this is the dimension? Is this attribute used in the analysis? I would guess it is not required...

File: pycqed/analysis_v2/Two_state_T1_analysis.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+
+    set_xlabel(ax, xlabel, xunit)
+    set_ylabel(ax, ylabel)
+    ax.grid(axis='y')
+    ax.legend()
+    ax.set_ylim(-.05, 1.05)
+    ax.set_title(title)
+
+
+def populations_using_rate_equations(SI: np.array, SX: np.array,
```
Comment (AdriaanRol) [inline pycqed/analysis_v2/Two_state_T1_analysis.py]: Please remove the function you have copy pasted and  replace it with an appropriate import statement at the top of the file.

File: pycqed/analysis_v2/Two_state_T1_analysis.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+            self.proc_data_dict['P1'][val_name] = P1
+            self.proc_data_dict['P2'][val_name] = P2
+            self.proc_data_dict['M_inv'][val_name] = M_inv
+
+    def run_fitting(self):
+        super().run_fitting()
+        self.fit_res['fit_res_P2'] = OrderedDict()
+        self.fit_res['fit_res_P1'] = OrderedDict()
+        self.fit_res['fit_res_P0'] = OrderedDict()
+        decay_mod = lmfit.Model(ExpDecayFunclocal, independent_vars='t')
```
Comment (AdriaanRol) [inline pycqed/analysis_v2/Two_state_T1_analysis.py]: Why is there a local copy of this (`ExpDecayFunclocal`) function? Is there anything different between this one and the one in fit_mods?

File: pycqed/analysis_v2/Two_state_T1_analysis.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+        decay_mod.set_param_hint('tau', value=15e-6, min=0, vary=True)
```
Comment (AdriaanRol) [inline pycqed/analysis_v2/Two_state_T1_analysis.py]: Most models have a `guess` method that gives guesses for these parameters. You do not have to do that now (this works) but keep it in mind for the future.

File: pycqed/analysis_v2/Two_state_T1_analysis.py [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+            tau1_best = fit_res_P1.best_values['tau1']
+            amplitude1_best = fit_res_P1.best_values['amplitude1']
+            amplitude2_best = fit_res_P1.best_values['amplitude2']
+            text_msg += ('\n' +
+                         r'$T_1^{eg}$  : ' + SI_val_to_msg_str
+                         (round(fit_res_P1.params['tau1'].value, 6), 's')[0]
+                         + SI_val_to_msg_str
+                         (fit_res_P1.params['tau1'].value, 's')[1])
+        except Exception as e:
+            logging.warning("Doulbe Fitting failed")
```
Comment (AdriaanRol) [inline pycqed/analysis_v2/Two_state_T1_analysis.py]: Typo: Doulbe Fitting failed-> Double Fitting failed

File: pycqed/analysis_v2/Two_state_T1_analysis.py [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+            logging.warning(e)
+            self.fit_res['fit_res_P0'] = {}
+
+        self.proc_data_dict['fit_msg'] = text_msg
+
+    def prepare_plots(self):
+        val_names = self.raw_data_dict['value_names']
+
```
Comment (AdriaanRol) [inline pycqed/analysis_v2/Two_state_T1_analysis.py]: Could you paste the figure that your test generates as a comment in this Pull Request? That helps assessing wether the plotting makes sense.

File: pycqed/analysis_v2/Two_state_T1_analysis.py [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+    ax.plot(time, P0, c='C0', linestyle='',
+            label=r'P($|g\rangle$)', marker='v')
+    ax.plot(time, P1, c='C1', linestyle='',
+            label=r'P($|e\rangle$)', marker='^')
+    ax.plot(time, P2, c='C2', linestyle='',
+            label=r'P($|f\rangle$)', marker='d')
+
+    set_xlabel(ax, xlabel, xunit)
+    set_ylabel(ax, ylabel)
+    ax.grid(axis='y')
```
Comment (AdriaanRol) [inline pycqed/analysis_v2/Two_state_T1_analysis.py]: please remove the grid here. In certain other contexts it has meaning but here it is only visual bloat.

File: pycqed/analysis_v2/Two_state_T1_analysis.py [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+        p0, p1 = np.dot(M_inv, np.array([sI-V2, sX-V2]))
+        P0[i] = p0
+        P1[i] = p1
+
+    P2 = 1 - P0 - P1
+
+    return P0, P1, P2, M_inv
+
+
+def ExpDecayFunclocal(t, tau, amplitude, offset):
```
Comment (AdriaanRol) [inline pycqed/analysis_v2/Two_state_T1_analysis.py]: Please look into existing fitting functions, creating a million double "local" functions is bad practice. Of course if your function is different it should be created, but I'm quite sure there are existing functions for this.

File: pycqed/tests/analysis_v2/test_Two_state_T1_analysis.py [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    @classmethod
+    def setUpClass(self):
+        self.datadir = os.path.join(pq.__path__[0], 'tests', 'test_data')
+        ma.a_tools.datadir = self.datadir
+
+    def test_efT1_analysis(self):
+        b = Ta.efT1_analysis(
+            t_start='20180606_144110', auto=True, close_figs=False)
+        t1 = b.fit_res['fit_res_P0'].params['tau1'].value
+        self.assertAlmostEqual(t1*1e6, 33.393, places=1)
```
Comment (AdriaanRol) [inline pycqed/tests/analysis_v2/test_Two_state_T1_analysis.py]: You are making multiple fits but are only testing one value, maybe also test another one?

File: pycqed/analysis_v2/Two_state_T1_analysis.py (hunk 7 above)
Comment (ByoungMooAnn) [reply pycqed/analysis_v2/Two_state_T1_analysis.py]: Okay.

File: pycqed/tests/analysis_v2/test_Two_state_T1_analysis.py (hunk 9 above)
Comment (ByoungMooAnn) [reply pycqed/tests/analysis_v2/test_Two_state_T1_analysis.py]: Okay.

File: pycqed/analysis_v2/Two_state_T1_analysis.py (hunk 4 above)
Comment (ByoungMooAnn) [reply pycqed/analysis_v2/Two_state_T1_analysis.py]: Okay.

File: pycqed/analysis_v2/Two_state_T1_analysis.py (hunk 6 above)
Comment (ByoungMooAnn) [reply pycqed/analysis_v2/Two_state_T1_analysis.py]: plot_populations_digitized w2_20180606_144110

File: pycqed/analysis_v2/Two_state_T1_analysis.py (hunk 2 above)
Comment (ByoungMooAnn) [reply pycqed/analysis_v2/Two_state_T1_analysis.py]: Okay.

File: pycqed/analysis_v2/Two_state_T1_analysis.py (hunk 1 above)
Comment (ByoungMooAnn) [reply pycqed/analysis_v2/Two_state_T1_analysis.py]: I have checked that removing 26 does not make difference.

File: pycqed/analysis_v2/Two_state_T1_analysis.py (hunk 3 above)
Comment (ByoungMooAnn) [reply pycqed/analysis_v2/Two_state_T1_analysis.py]: Just for my convenience but I have changed. Now code is using fitting_model module.

File: pycqed/analysis_v2/Two_state_T1_analysis.py (hunk 5 above)
Comment (ByoungMooAnn) [reply pycqed/analysis_v2/Two_state_T1_analysis.py]: Okay.

**Pull request merged**